README.md
# NAME

Term::Chrome - DSL for colors and other terminal chrome

# SYNOPSIS

    use Term::Chrome qw<Red Blue Bold Reset Underline Green color>;

    # Base color constant and attribute
    say Red, 'red text', Reset;

    # Composition, using operator overloading
    say Red/Blue+Bold, 'red on blue', Reset;

    # Undo
    say Bold, 'bold', !Bold, 'not bold';

    # Extended xterm-256 colors
    say color(125) + Underline, 'Purple', Reset;

    # Define your own constants
    use constant Pink => color 213;

    # Use ${} around Chrome expression inside strings
    say "normal ${ Red+Bold } RED ${ +Reset } normal";

    # Extract components
    say( (Red/Blue)->bg, "blue text", (Green+Reset)->flags );

Chromizer: get a closure that applies given chrome before, and undo after
the argument.

    # Get an efficient chromizer
    my $boldifier = \&{ +Bold };
    # Use the chromizer
    say $boldifier->("bold text");
    # Same as:
    say Bold, "bold text", !Bold;

    # Short lived chromizers using color literals:
    say(&{+Red}('red'));
    say(&{ Red/Blue + Bold }('red on blue'));
    # Same, but requires perl 5.21.4+
    #say(( Red/Blue + Bold )->('red on blue'));

# DESCRIPTION

`Term::Chrome` is a domain-specific language (DSL) for terminal decoration
(colors and other attributes).

In the current implementation stringification to ANSI sequences for `xterm`
and `xterm-256` is hard-coded (which means it doesn't use the [terminfo(5)](http://man.he.net/man5/terminfo)
database), but this gives optimized (short) strings.

Colors and attributes are exposed as objects that have overloading for
arithmetic operators.

# EXPORTS

## Functions

`color(_0-255_)`

Build a [Term::Chrome](https://metacpan.org/pod/Term::Chrome) object with the given color number. You can use this
constructor to create your own set of color constants.

For example, `color(0)` gives the same result as `Black` (but not the same
object).

## Colors

Each of these function return a Chrome object.

- `Black`: `color 0`
- `Red`: `color 1`
- `Green`: `color 2`
- `Yellow`: `color 3`
- `Blue`: `color 4`
- `Magenta`: `color 5`
- `Cyan`: `color 6`
- `White`: `color 7`

## Decoration flags

The exact rendering of each flag is dependent on how the terminal implements
them. For example `Underline` and `Blink` may do nothing.

- `Bold`
- `Underline`
- `Blink`
- `Reverse`

## Special flags

- `Reset`: reset all colors and flags
- `ResetFlags`: reset (undo) all chrome flags (Bold, Underline, Blink, Reverse)
- `ResetFg`: reset (undo) foreground color
- `ResetBg`: reset (undo) background color

# METHODS

Here are the methods on `Term::Chrome` objects:

- `fg`

    Extract the Chrome object of just the foreground color. Maybe `undef`.

- `bg`

    Extract the Chrome object of the just background color. Maybe `undef`.

- `flags`

    Extract a Chrome object of just the decoration flags. Maybe `undef`.

# OVERLOADED OPERATORS

- `/` (mnemonic: "over")

    Combine a foreground color (on the left) with a background color.

    Example:

        my $red_on_black = Red / Black;

- `+`

    Add decoration flags (on the right) to colors (on the left).

    Example:

        my $bold_red = Red + Bold;

- `!` (negation)

    Returns a chrome which is the reverse of chrome to which negation is applied.

        my $reset_foreground = ! Red;
        my $reset_colors = ! (Red / Black);

    The reverse of `Reset`, `ResetFg`, `ResetBg`, `ResetFlags` is nothing.

- `""` (stringification)

    Transform the object into a string of ANSI sequences. This is
    particularly useful to directly use a Chrome object in a double quoted string.

- `${}` (scalar dereference)

    Same result as `""` (stringification). This operator is overloaded because
    it is convenient to interpolate Chrome expressions in double-quoted strings.

    Examples:

        say "normal ${ +Red } red ${ +Reset }";
        say "normal ${ Red + Bold } red ${ +Reset }";

    Note that you must force expression parsing context when a Chrome constant is
    used alone inside `${ }`: `${ +Red }` or `${ (Red) }` or `${ Red() }`.
    `use strict 'vars';` will detect those cases, but you may miss them in
    one-liners.

- `&{}` (code dereference, or _codulation_)

    Wrap some text with the given chrome and `Reset`.

    Example:

        say Red->("red text");
        # Same result as:
        say Red, "red text", Reset;

    Unfortunately perl had a bug
    ([perl RT#122607](https://rt.perl.org/Ticket/Display.html?id=122607)) that makes this feature not much usable in practice when applied to constants. That bug
    is fixed in perl 5.21.4+.
    On perl < 5.21.4 you have to wrap the chrome constant in a `do {}` or use `&{}()`:

        say do{ Red }->("red text");
        say &{ +Red }("red text");

    Codulation can also be used to extract a colorizer sub that will be more
    efficient if you use it multiple times:

        my $redifier = \&{ Red };
        say $redifier->("red text");

# BUGS

See the warnings about `${}` and `&{}` above.

# REFERENCES

- ISO 6429 / ECMA 48: https://www.ecma-international.org/publications/files/ECMA-ST/Ecma-048.pdf (reference for SGR is at page 61, numbered 75 in PDF)
- XTerm Control Sequences: http://invisible-island.net/xterm/ctlseqs/ctlseqs.html#h2-Functions-using-CSI-\_-ordered-by-the-final-character\_s\_

# SEE ALSO

Comments on each modules are opinions of the author.

- [Term::ANSIColor](https://metacpan.org/pod/Term::ANSIColor): the same basic features (and the others should not be in
Term::ANSIColor itself but in an extension), but with an awful API I could never
even consider to use while keeping my sanity.
- [Term::ScreenColor](https://metacpan.org/pod/Term::ScreenColor)
- [PerlIO::via::ANSIColor](https://metacpan.org/pod/PerlIO::via::ANSIColor)
- [AngelPS1](https://metacpan.org/pod/AngelPS1) or [https://github.com/dolmen/angel-PS1](https://github.com/dolmen/angel-PS1): "The Angel's Prompt" is
the project for which `Term::Chrome` has been built. [AngelPS1::Compiler](https://metacpan.org/pod/AngelPS1::Compiler),
the `angel-PS1` compiler has special support for `Term::Chrome` values.

# TRIVIA

Did you know that _chartreuse_ is one of the favorite colors of Larry Wall?

# AUTHOR

Olivier Mengué <dolmen@cpan.org>

# COPYRIGHT & LICENSE

Copyright © 2013-2018 Olivier Mengué.

This library is free software; you can redistribute it and/or modify it under
the same terms as Perl 5 itself.


Changes
Revision history for Term-Chrome

{{$NEXT}}

2.01  2018-02-20 23:40:44 CET
	- Fix reverse of Reset and attribute flags
	- Improve documentation of Reset flags
	- Repository moved to github.com/dolmen-perl5/Term-Chrome

2.00  2016-03-04 01:17:31 CET
	- Add '!' operator overloading to reverse a chrome:
           - !Bold means 'unbold',
           - !Reset is a no-op,
	   - !Red means 'reset foreground color',
           - !(Red/Blue) means 'reset foreground and background colors'
	- Chromizers: after the text, reverse the chrome instead of
	  applying a full "Reset" of the chrome. This is a BREAKING CHANGE
	  as the stringification is changing, but it is more DWIM.
	- Add 'bool' operator overloading. This allows to use '||' to set a
	  default value:   $chrome ||= Red;
	  (this was already working without the explicit overloading, except
	  under Devel::Cover, see t/50-cover-bug.t)

1.013  2016-02-17 21:58:41 CET
	- Improved doc for ${} and &{} overloading
	- Test::Synopsis is now an *optional* test dependency
	- Dédicace spéciale à Andéol !

1.012  2015-11-03 01:00:13 CET
	- Document a workaround for codulation of constants on perl < 5.21.4
	  (use a do {} around the constant)

1.011  2015-10-21 09:13:44 CEST
	- Fix t/12-codulation.t for perl >= 5.21.4

1.010  2015-10-21 01:51:21 CEST
	- Add &{} overloading ("codulate") to easily wrap some text with color
	  and Reset. Unfortunately, it is not usable in practice due a core
	  perl bug < 5.21.4.
	  This allows to build efficient "chromizers" (closures that colorize
	  and Reset).
	- Disable fallbacks for overload.pm, to reduce (dangerous) magic
	- Remove ->clone method that is useless as Term::Chrome values are
	  immutables

1.001  2014-08-24 00:43:54 EEST
	- POD: add SEE ALSO links

1.000  2014-08-23 12:35:33 EEST
        - Initial release, from Sofia, at YAPC::EU::2014
	  (Code is mature, extracted from angel-PS1)


benchmark.pl
#!/usr/bin/env perl

use strict;
use warnings;

use Benchmark ':all';

use lib 'lib';
use Term::Chrome;
use Term::ANSIColor qw(colored BOLD RED ON_BLUE RESET YELLOW ON_MAGENTA);

my $s = 'a' x 50;

my %bench = (
    'Chrome normal' => sub {
	join('', Yellow / Magenta, "Yellow on magenta $s", Reset)
    },
    'Chrome CODEREF' => sub {
	&{ Yellow / Magenta }("Yellow on magenta $s")
    },
    'Chrome cached' => do {
	my $yellow_on_magenta = \&{ Yellow / Magenta };
	sub {
	    $yellow_on_magenta->("Yellow on magenta $s")
	}
    },
    'ANSIColor colored' => sub {
	colored(['yellow on_magenta'], "Yellow on magenta $s")
    },
    'ANSIColor constants' => sub {
	join('', YELLOW ON_MAGENTA "Yellow on magenta $s", RESET)
    },
    'ANSIColor autoreset' => sub {
	local $Term::ANSIColor::AUTORESET = 1;
	YELLOW ON_MAGENTA "Yellow on magenta $s"
    },
);

my $Redifier = \&{+Red};
sub dump_bench
{
    for my $name (sort keys %bench) {
	my $result = $bench{$name}->();
	$result =~ s/\e(\[.*?[a-zA-Z])/$Redifier->("\\e$1")/ge;
	printf "%s:\n%s\n", $name, $result
    }
}

dump_bench;

cmpthese(2000000, \%bench);

%bench = (
    'Chrome constants' => sub {
	join('', Red / Blue + Bold, "Bold red on blue.", Reset)
    },
    'ANSIColor constants' => sub {
	join('', RED BOLD ON_BLUE "Bold red on blue.", RESET)
    },
);
dump_bench;

cmpthese(2000000, \%bench);



lib/Term/Chrome.pm
use strict;
use warnings;

package Term::Chrome;
# ABSTRACT: DSL for colors and other terminal chrome
our $VERSION = '2.00';

# Pre-declare packages
{
    package # no index: private package
        Term::Chrome::Color;
}


use Exporter 5.57 'import';  # perl 5.8.3
# @EXPORT is defined at the end

use Carp ();
use Scalar::Util ();
our @CARP_NOT = qw< Term::Chrome::Color >;

# Private constructor for Term::Chrome objects. Lexical, so cross-packages.
# Arguments:
# - class name
# - foreground color
# - background color
# - flags list
my $new = sub
{
    my ($class, @self) = @_;

    my $fg = $self[0];
    Carp::croak "invalid fg color $fg"
        if defined($fg) && ($fg < 0 || $fg > 255);
    my $bg = $self[1];
    Carp::croak "invalid bg color $bg"
        if defined($bg) && ($bg < 0 || $bg > 255);
    # TODO check flags

    bless \@self, $class
};


# Cache for color objects
my %COLOR_CACHE;

sub color ($)
{
    my $color = shift;
    die "invalid color" if ref $color;
    my $c = chr $color;
    # We can not use '$COLOR_CACHE{$c} ||= ...' because this requires overloading
    # We can not use 'no overloading' because this requires perl 5.10
    exists $COLOR_CACHE{$c}
    ?  $COLOR_CACHE{$c}
    : ($COLOR_CACHE{$c} = Term::Chrome::Color->$new($color, undef))
}


use overload
    '""' => 'term',
    '+'  => '_plus',
    '${}' => '_deref',
    '&{}' => '_chromizer',
    '.'   => '_concat',
    '!'   => '_reverse',
    'bool' => sub () { 1 },
    fallback => 0,
;

sub term
{
    my $self = shift;
    my ($fg, $bg) = @{$self}[0, 1];
    my $r = join(';', @{$self}[2 .. $#$self]);
    if (defined($fg) || defined($bg)) {
        $r .= ';' if @$self > 2;
        if (defined $fg) {
            # LeoNerd says that this should be ----------> "38:5:$fg"
            # according to the spec but gnome-terminal doesn't support that
            $r .= $fg < 8 ? (30+$fg) : $fg < 16 ? "9$fg" : "38;5;$fg";
            $r .= ';' if defined $bg;
        }
        #                                      -------> "48:5:$bg"
        $r .= $bg < 8 ? (40+$bg) : $bg < 16 ? "10$bg" : "48;5;$bg" if defined $bg;
    } else {
        return '' unless @$self > 2
    }
    "\e[${r}m"
}


sub _plus
{
    my ($self, $other, $swap) = @_;

    return $self unless defined $other;

    die 'invalid value for +' unless $other->isa(__PACKAGE__);

    my @new = @$self;
    $new[0] = $other->[0] if defined $other->[0];
    $new[1] = $other->[1] if defined $other->[1];
    push @new, @{$other}[2 .. $#$other];

    bless \@new
}

my %reverse = (
    # Unfortunately there isn't a perfect mapping
    # Reference:
    # https://www.ecma-international.org/publications/files/ECMA-ST/Ecma-048.pdf page 75
    1 => 22,
    2 => 22,
    3 => 23,
    4 => 24,   # Underlined
    5 => 25,
    6 => 25,
    7 => 27,
    8 => 28,
    9 => 29,
    21 => 24,  # Double underline

    22 => 1,
    23 => 3,
    24 => 4,
    25 => 5,
    27 => 7,
    28 => 8,
    29 => 9,
);

sub _reverse
{
    my $self = shift;
    my @new = (undef, undef);
    push @new, 39 if $self->[0]; # ResetFg
    push @new, 49 if $self->[1]; # ResetBg
    # Reset/ResetFlags/ResetFg/ResetBg are removed
    # Other flags are reversed
    push @new, map { (!$_ || $_ > 30 || !exists $reverse{$_}) ? () : $reverse{$_} } @{$self}[2..$#$self];
    bless \@new, 'Term::Chrome::Flag'
}

sub _deref
{
    \("$_[0]")
}

sub _concat
{
    $_[2] ? $_[1].$_[0]->term
          : $_[0]->term.$_[1]
}


sub _chromizer
{
    my $self = shift;
    my $begin = $self->term;
    my $end = $self->_reverse->term;
    sub {
        unless (defined $_[0]) {
            Carp::carp "missing argument in Term::Chrome chromizer";
            return
        }
        $begin . $_[0] . $end
    }
}

sub fg
{
    my $c = $_[0]->[0];
    defined($c) ? color($c) : undef
}

sub bg
{
    my $c = $_[0]->[1];
    defined($c) ? color($c) : undef
}

sub flags
{
    my $self = shift;
    return undef unless @$self > 2;
    __PACKAGE__->$new(undef, undef, @{$self}[2..$#$self])
}

package # no index: private package
    Term::Chrome::Color;

our @ISA = qw< Term::Chrome >;

use overload
    '/'   => '_over',
    # Even if overloading is set in the super class, we have to repeat it for old perls
    (
        $^V ge v5.18.0
        ? ()
        : (
            '""'  => \&Term::Chrome::term,
            '+'   => \&Term::Chrome::_plus,
            '${}' => \&Term::Chrome::_deref,
            '.'   => \&Term::Chrome::_concat,
            '!'   => \&Term::Chrome::_reverse,
            'bool' => sub () { 1 },
        )
    ),
    fallback => 0,
;

sub _over
{
    die 'invalid bg color for /' unless ref($_[1]) eq __PACKAGE__;
    Term::Chrome->$new($_[0]->[0], $_[1]->[0])
}

package # no index: private package
    Term::Chrome::Flag;

our @ISA = qw< Term::Chrome >;

use overload
    '+'   => '_plus',
    '!'   => '_reverse',
    # Even if overloading is set in the super class, we have to repeat it for old perls
    (
        $^V ge v5.18.0
        ? ()
        : (
            '""'  => \&Term::Chrome::term,
            '${}' => \&Term::Chrome::_deref,
            '.'   => \&Term::Chrome::_concat,
            'bool' => sub () { 1 },
        )
    ),
    fallback => 0,
;

sub _reverse
{
    my $self = shift;
    bless [
        undef, undef,
        # Reset/ResetFlags/ResetFg/ResetBg are removed
        map { (!$_ || $_ > 30 || !exists $reverse{$_}) ? () : $reverse{$_} } @{$self}[2..$#$self]
    ]
}

sub _plus
{
    my ($self, $other, $swap) = @_;

    return $self unless defined $other;

    Carp::croak(q{Can't combine Term::Chrome with }.$other)
        unless Scalar::Util::blessed $other;

    if ($other->isa(__PACKAGE__)) {
        # Reset
        return $other if !$other->[2];
        # ResetFlags
        return $other if $#$other == 8 || ($self->[2] && $self->[2] < 30 && $other->[2] == $reverse{$self->[2]});
        # Concat flags
        __PACKAGE__->$new(@$self, @{$other}[2..$#$other])
    } elsif ($other->isa(Term::Chrome::)) {
        $other->_plus($self, '')
    } else {
        Carp::croak(q{Can't combine Term::Chrome with }.ref($other))
    }
}


package
    Term::Chrome;

# Build the constants and the @EXPORT list
#
# This block must be after "use overload" (for both Term::Chrome
# and Term::Chrome::Color) because overload must be set before blessing
# due to a bug in perl < 5.18
# (according to a comment in Types::Serialiser source)

my $mk_flag = sub { Term::Chrome::Flag->$new(undef, undef, @_) };

my %const = (
    Reset      => $mk_flag->(''),
    ResetFg    => $mk_flag->(39),
    ResetBg    => $mk_flag->(49),
    ResetFlags => $mk_flag->(22, 23, 24, 25, 27, 28),
    Standout   => $mk_flag->(7),
    Underline  => $mk_flag->(4),
    Reverse    => $mk_flag->(7),
    Blink      => $mk_flag->(5),
    Bold       => $mk_flag->(1),

    Black      => color 0,
    Red        => color 1,
    Green      => color 2,
    Yellow     => color 3,
    Blue       => color 4,
    Magenta    => color 5,
    Cyan       => color 6,
    White      => color 7,

    # Larry Wall's favorite color
    # The true 'chartreuse' color from X11 colors is #7fff00
    # The xterm-256 color #118 is near: #87ff00
    Chartreuse => color 118,
);

our @EXPORT = ('color', keys %const);

# In 17fd029f we avoided to use constant.pm on perl < 5.16
# This does not seem necessary anymore.
require constant;
constant->import(\%const);

1;
# vim:set et ts=8 sw=4 sts=4:
